homework7/hw7task2/project2_/task2_backspace_compare.py

```diff
@@ -0,0 +1,42 @@
+"""
+Given two strings. Return if they are equal when both are typed into
+empty text editors. # means a backspace character.
+Note that after backspacing an empty text, the text will continue empty.
+Examples:
+    Input: s = "ab#c", t = "ad#c"
+    Output: True
+    # Both s and t become "ac".
+    Input: s = "a##c", t = "#a#c"
+    Output: True
+    Explanation: Both s and t become "c".
+    Input: a = "a#c", t = "b"
+    Output: False
+    Explanation: s becomes "c" while t becomes "b".
+"""
+
+
+def handle_with_backspaces(string: str) -> list:
+    """One backspace(#) deletes one symbol which is before him in string.
+
+    Return a list of symbols which were not deleted by backspaces.
+    """
+
+    backspace = "#"
+    backspaces_amount = string.count(backspace)
+    symbols = list(string)
+
+    for i in range(backspaces_amount):
+        index_backspace = symbols.index(backspace)
+        del symbols[index_backspace]
+        if index_backspace != 0:
+            del symbols[index_backspace - 1]
+
+    return symbols
```

Comment:
Could you describe the complexity of this solution for each line?

Reply:
`backspaces_amount = string.count(backspace) ` **O(n)** , where n= len(string)
`symbols = list(string)` O(n), as list('abc') return ['a', 'b', 'c']
`for i in range(backspaces_amount):` **O(backspaces_amount)** where the worst case is **n**, where all elements in string are backspaces
`index_backspace = symbols.index(backspace)` **O(n)**
`del symbols[index_backspace]` **O(n)**
`if index_backspace != 0:`   **O(1)**
`del symbols[index_backspace - 1]`  **O(n)**

So, total O = O(n) + O(n) + O(n) * (O(n)+O(n) +  (O(1)* O(n))  )= O(2*n) + O(n)* O(3*n) ~= O(3*n**2)

Reply:
I agree with that, here are some notes about simplification:


`O(n) + O(n) + O(b) * ( O(n) + O(n) + O(1) + O(n))`

Simplification rules:   we drop coefficients.
```
O(n) + O(n) -> O(2n) -> O(n)
```

The result is `O(n) + O(b) * O(n)`   Since b can be n sometimes we could replace O(b) with O(n)

Second simplification rule: function that grows faster suppress another one, we can drop slow one
```
O(n) + O(n^2)  -> O(n^2) 
```

The final result is `O(n^2)`.  So let's improve it to O(n).  Just remember how you are doing computation of this manually, do you read the string from the beginning multiple times?

Reply:
Done. 
    `for s in symbols:`            **O(n)**
        `if s == backspace:`     **O(1)**
           ` if len(stack): `          **O(1)**  
                `stack.pop()`        **O(1)** 
        `else:`
            `stack.append(s)`     **O(1)** 
Total = O(n)